### ISSUE TITLE
Extend TransactionInputInterface and TransactionOutputInterface ?

### ISSUE THREAD

**Issue by aztech-dev** [afk11/bitcoin#9]
While refactoring for #7, I noticed that there are interfaces for transaction inputs and outputs, but the library is typed hinted against the implementations.

I'm assuming that's because the implementations have extra methods that the interfaces do not (the interfaces are readonly). The collection classes I added are typed against the interfaces, which is better IMO, but some code depends on some methods that the interfaces do not have (all the setter methods).

There are multiple ways to solve this, but the main two are:

- Add the missing method to the interfaces. 
- Add new WritableInputInterface and WritableOutputInterface interfaces, that extend the original interfaces, and have the implementations implement those as well. This solution is cleaner IMO, as it segregates readonly and writeable transactions, but it will require some typechecks at some points to ensure that inputs/outputs returned by the collections are writable. Or better yet, add new Writeable collections... Oh wait, better yet, make the Transaction class completely readonly, and introduce a new TransactionBuilder class (I pondered for a second about having a writable transaction class, but I think it would bring a lot of headaches considering PHP's class system).

**Comment by afk11** [comment]
Yeah not sure what I did there, but yeah it's best to typehint the interfaces :) Bitcoins reference implementation also has mutable and immutable transaction classes, though not sure of how it uses them differently. 

A transaction builder in general is a good idea I think, it would keep the logic signing transactions to outside of the main Transaction class (similar to the Signer for producing signatures ;) )

**Comment by afk11** [comment]
I've started writing a TransactionBuilder now. Starting to look at this issue more as I'm trying to maintain all the input scripts and the state that goes with it. Thinking those scripts should come from a SignatureCollection that the builder maintains for each input, and serializes according to the inputs prevOutput() script type. 

(There is presently requirement if signing a TransactionInput, that the TransactionOutput that it spends must be set) 

And if simply verifying a transaction normal you would just parse the scripts directly.
